### PR TITLE
Material allows "select all" when not collapsed

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Problem
Currently, `TextSelectionControls` in both cupertino and material disallows "select all" when selection is not collapsed.

However, Android actually allows "select all" even when the selection is not collapsed, unless the selection is already "all". (The current behavior matches the one on iOS.)

The following example shows that Android allows "Select All" on partial selection (right) while Flutter doesn't (left)
image

The following example shows that Android disallows "Select All" on full selection
image

On iOS it's correct.
image

## Changes

This PR aligns `_MaterialTextSelectionControls` with the Android behavior by overriding `canSelectAll`. Now `materialTextSelectionControls` allows "select all" when text is partially selected.

## Related Issues
None

## Tests

I added the following tests:
- For each of `{material,cupertino}TextSelectionControls`, test the result of `canSelectAll` under the following circumstances respectively:
  - No text
  - Text and collapsed selection
  - Text and partial selection
  - Text and full selection

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/cupertino/text_selection_test.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter_test/flutter_test.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter/cupertino.dart';
+
+void main() {
```
Comment (justinmc) [inline packages/flutter/test/cupertino/text_selection_test.dart]: Thanks for creating these two test files 👍

File: packages/flutter/lib/src/material/text_selection.dart [hunk 2]
```diff
@@ -212,6 +212,15 @@ class _MaterialTextSelectionControls extends TextSelectionControls {
     assert(type != null);
     return null;
   }
+
+  @override
+  bool canSelectAll(TextSelectionDelegate delegate) {
+    // Android allows SelectAll when selection is not collapsed, unless all
+    // have already been selected.
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_selection.dart]: Maybe this is more clear, but up to you:

"unless everything has already been selected"

**Review by justinmc — APPROVED** [review]
LGTM, thanks for making the description so clear with screenshots!  I definitely agree that we should fix this.

I think the failures might be legitimate cases where this has broken the golden files, you should definitely watch out for that.  For example, I see this one failing: https://github.com/flutter/goldens/blob/master/packages/flutter/test/material/text_field_opacity_test.0.1.png

File: packages/flutter/test/cupertino/text_selection_test.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter/test/cupertino/text_selection_test.dart]: Thank you for review!

File: packages/flutter/lib/src/material/text_selection.dart (hunk 2 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/text_selection.dart]: Fixed

**Pull request merged**